lib/alloc.c

```diff
@@ -71,3 +71,23 @@ xreallocarray(void *p, size_t nmemb, size_t size)
 	        log_get_progname(), strerror(errno));
 	exit(13);
 }
+
+char *
+xstrndup(const char *s, size_t max_length) {
+    size_t length = strlen(s);
+    if (length > max_length) {
+        length = max_length;
+    }
+
+    char *p = malloc(length + 1);
```

Comment:
We don't call malloc(3) directly.  Have a look at <lib/alloc.h>, and the MALLOC() macro.

Reply:
And you'll also want to check XMALLOC().

Reply:
New approach is applied by direct usage of a new function: is_valid_user_name